src/operator/nn/mkldnn/mkldnn_rnn_impl.h

```diff
@@ -0,0 +1,589 @@
+  std::vector<primitive::at> inputs;
+  for (size_t i = 0; i < srcs_cds.size(); i++) {
+    inputs.push_back(srcs[i]);
+  }
+
+  auto dst_desc = mkldnn::memory::desc(dst_cds, mkldnn_dtype, dst_format);
+  auto concat_pd = mkldnn::concat::primitive_desc(dst_desc, concat_dimension, srcs_pd);
+  auto dst = mkldnn::memory(concat_pd.dst_primitive_desc());
+  
+  auto c = mkldnn::concat(concat_pd, inputs, dst);
+  stream(stream::kind::eager).submit({c}).wait();
+  return dst;
+}
+
+template <typename DType>
+void MKLDNNRNNForward(bool state_outputs,
+                      int L,
+                      int D,
+                      const int T,
+                      const int N,
+                      int I,
+                      const int H,
+                      DType* x_ptr,
+                      DType* hx_ptr,
+                      DType* cx_ptr,
+                      DType* w_ptr,
+                      DType* b_ptr,
+                      DType* y_ptr,
+                      DType* hy_ptr,
+                      DType* cy_ptr,
+                      std::vector<mkldnn::memory> *weight_bias_memory,
+                      int mode) {
+  int ngates = 0, nstates = 0;
+  mkldnn::algorithm nalgorithm = GetMKLDNNAlgo(mode, &ngates, &nstates);
+  mkldnn::memory::data_type mkldnn_dtype = GetMKLDNNDataType<DType>();
+  const int b_size = 2 * H * ngates * D;
+  const int cell_size = N * H * D;
+  const int single_cell_size = N * H;
+  const int single_b_size = ngates * H;
+  //  First layer
+  int w_size = (I + H) * H * ngates * D;
+  DType* x_0 = x_ptr;  //  T * N * I
+  DType* y = y_ptr;  //  T * N * D * H
+  DType* back_w_ptr = w_ptr;
+  DType* back_b_ptr = b_ptr;
+  DType* wx_0 = w_ptr;  //  ngates * H, I
+  DType* wh_0 = w_ptr + I * H * ngates;  //  ngates * H, H
+  if (D == 2) {
+    back_w_ptr = w_ptr + ngates * H * (I + H);
+    back_b_ptr = b_ptr + single_b_size * 2;
+  }
+  DType* back_wx_0 = back_w_ptr;
+  DType* back_wh_0 = back_w_ptr + I * H * ngates;
+  DType* bx_0 = b_ptr;
+  DType* bh_0 = b_ptr + H * ngates;
+  DType* back_bx_0 = back_b_ptr;
+  DType* back_bh_0 = back_b_ptr + H * ngates;
+  const int omp_threads = mxnet::engine::OpenMP::Get()->GetRecommendedOMPThreadCount();
+  using namespace mkldnn;
+  auto cpu_engine = CpuEngine::Get()->get_engine();
+  auto null_memory_ = null_memory(cpu_engine);
+  std::vector<primitive> rnn_net_0;
+  memory::dims src_layer_tz_0 = {T, N, I};
+  memory::dims weights_layer_tz_0 = {1, D, I, ngates, H};  //  ldigo
+  memory::dims weights_layer_r_tz_0 = {1, 1, I, ngates, H};  //  ldigo for reorder
+  memory::dims weights_iter_tz_0 = {1, D, H, ngates, H};  //  ldigo
+  memory::dims weights_iter_r_tz_0 = {1, 1, H, ngates, H};  //  ldigo for reorder
+  memory::dims bias_tz_0 = {1, D, ngates, H};
+  memory::dims dst_layer_tz_0 = {T, N, D * H};
+  memory::dims src_iter_tz_0 = {1, D, nstates, N, H};  //  ldsnc
+  memory::dims dst_iter_tz_0 = {1, D, nstates, N, H};  //  ldsnc
+  int offset1 = 0, offset2 = 0;
```

Comment:
minor comments for the variables naming, please use more meaningful names. maybe change `back_xxx` to `l2r_xxx` or `r2l_xxx`?